pyomo/gdp/plugins/binary_multiplication.py

```diff
@@ -0,0 +1,175 @@
+from pyomo.core.util import target_list
+from weakref import ref as weakref_ref
+import logging
+
+
+logger = logging.getLogger('pyomo.gdp.binary_multiplication')
+
+
+@TransformationFactory.register(
+    'gdp.binary_multiplication',
+    doc="Reformulate the GDP as an MINLP by multiplying f(x) <= 0 by y to get f(x) * y <= 0.",
+)
+class GDPToMINLPTransformation(GDP_to_MIP_Transformation):
+    CONFIG = ConfigDict("gdp.binary_multiplication")
+    CONFIG.declare(
+        'targets',
+        ConfigValue(
+            default=None,
+            domain=target_list,
+            description="target or list of targets that will be relaxed",
+            doc="""
+
+        This specifies the list of components to relax. If None (default), the
+        entire model is transformed. Note that if the transformation is done out
+        of place, the list of targets should be attached to the model before it
+        is cloned, and the list will specify the targets on the cloned
+        instance.""",
+        ),
+    )
+
+    transformation_name = 'binary_multiplication'
+
+    def __init__(self):
+        super().__init__(logger)
+
+    def _apply_to(self, instance, **kwds):
+        try:
+            self._apply_to_impl(instance, **kwds)
+        finally:
+            self._restore_state()
+
+    def _apply_to_impl(self, instance, **kwds):
+        self._process_arguments(instance, **kwds)
+
+        # filter out inactive targets and handle case where targets aren't
+        # specified.
+        targets = self._filter_targets(instance)
+        # transform logical constraints based on targets
+        self._transform_logical_constraints(instance, targets)
+        # we need to preprocess targets to make sure that if there are any
+        # disjunctions in targets that their disjuncts appear before them in
+        # the list.
+        gdp_tree = self._get_gdp_tree_from_targets(instance, targets)
+        preprocessed_targets = gdp_tree.reverse_topological_sort()
+
+        for t in preprocessed_targets:
+            if t.ctype is Disjunction:
+                self._transform_disjunctionData(
+                    t,
+                    t.index(),
+                    parent_disjunct=gdp_tree.parent(t),
+                    root_disjunct=gdp_tree.root_disjunct(t),
+                )
+
+    def _transform_disjunctionData(
+        self, obj, index, parent_disjunct=None, root_disjunct=None
+    ):
+        (transBlock, xorConstraint) = self._setup_transform_disjunctionData(
+            obj, root_disjunct
+        )
+
+        # add or (or xor) constraint
+        or_expr = 0
+        for disjunct in obj.disjuncts:
+            or_expr += disjunct.binary_indicator_var
+            self._transform_disjunct(disjunct, transBlock)
+
+        rhs = 1 if parent_disjunct is None else parent_disjunct.binary_indicator_var
+        if obj.xor:
+            xorConstraint[index] = or_expr == rhs
```

Comment:
Unfortunately (and this is my fault too), this isn't actually correct... It assumes that nested indicator variables are local, which is not necessarily true. The safest thing for now is to always set the rhs to 1 and to put the exactly-one constraint on the parent block (and transform from leaf to root, which I think you're already doing).

Reply:
Your point with putting the exactly-one constraint on the parent block (instead of the root disjunct's parent block) is that the exactly-one constraint needs to be transformed again if it is for a nested disjunction?

Reply:
Maybe I'm misunderstanding what the "root disjunct" is...

Reply:
Is putting the exactly-one constraint on the parent disjunct equivalent?

Reply:
Correct. It should be on the parent disjunct so that it does get transformed again.